include/envoy/stats/stats.h

```diff
@@ -20,78 +20,77 @@ class Instance;
 namespace Stats {
 
 /**
+ * General interface for all stats objects.
+ */
+class Metric {
+public:
+  virtual ~Metric() {}
+  /**
+   * Returns the full name of the Metric.
+   */
+  virtual const std::string& name() const PURE;
+};
+
+/**
  * An always incrementing counter with latching capability. Each increment is added both to a
  * global counter as well as periodic counter. Calling latch() returns the periodic counter and
  * clears it.
  */
-class Counter {
+class Counter : public virtual Metric {
 public:
   virtual ~Counter() {}
   virtual void add(uint64_t amount) PURE;
   virtual void inc() PURE;
   virtual uint64_t latch() PURE;
-  virtual std::string name() PURE;
   virtual void reset() PURE;
-  virtual bool used() PURE;
-  virtual uint64_t value() PURE;
+  virtual bool used() const PURE;
+  virtual uint64_t value() const PURE;
 };
 
 typedef std::shared_ptr<Counter> CounterSharedPtr;
 
 /**
  * A gauge that can both increment and decrement.
  */
-class Gauge {
+class Gauge : public virtual Metric {
 public:
   virtual ~Gauge() {}
 
   virtual void add(uint64_t amount) PURE;
   virtual void dec() PURE;
   virtual void inc() PURE;
-  virtual std::string name() PURE;
   virtual void set(uint64_t value) PURE;
   virtual void sub(uint64_t amount) PURE;
-  virtual bool used() PURE;
-  virtual uint64_t value() PURE;
+  virtual bool used() const PURE;
+  virtual uint64_t value() const PURE;
 };
 
 typedef std::shared_ptr<Gauge> GaugeSharedPtr;
 
 /**
- * An individual timespan that is owned by a timer. The initial time is captured on construction.
- * A timespan must be completed via complete() for it to be stored. If the timespan is deleted
- * this will be treated as a cancellation.
+ * A histogram that records values one at a time.
  */
-class Timespan {
+class Histogram : public virtual Metric {
 public:
-  virtual ~Timespan() {}
+  virtual ~Histogram() {}
+
+  enum ValueType {
+    Integer,
+    Duration,
```

Comment:
What value does this subtyping provide? Would it be possible to treat everything as just integers at the interface level (and for time, it's just # of ms in the implementation already).

Reply:
FWIW this would be fine for Lyft, my concern is for implementation where somehow a histogram is different from a timer. I'm not sure if that is a real issue or not... Let's discuss tomorrow.